File: tests/cuda/lua/test_signal_acoustics.lua
-- Signal acoustics CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_acoustics.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

ins.set_device(ins.GPUPlace(0))

describe("Signal Acoustics CUDA Tests", function()
  it("mel2hz", function()
    local hz = ins.signal.mel2hz(1000.0)
    assert.is_not_nil(hz)
    assert.is_true(hz > 0)
  end)

  it("hz2mel", function()
    local mel = ins.signal.hz2mel(1000.0)
    assert.is_not_nil(mel)
    assert.is_true(mel > 0)
  end)

  it("mel_hz_roundtrip", function()
    local hz_orig = 1000.0
    local mel = ins.signal.hz2mel(hz_orig)
    local hz_back = ins.signal.mel2hz(mel)
    assert.is_true(math.abs(hz_back - hz_orig) < 1.0)
  end)

  it("mel_frequencies", function()
    local freqs = ins.signal.mel_frequencies(128, 0.0, 11025.0, false)
    assert.is_not_nil(freqs)
    assert.are.equal(128, freqs.numel)
  end)

  it("mel_frequencies_custom", function()
    local freqs = ins.signal.mel_frequencies(64, 20.0, 8000.0, false)
    assert.is_not_nil(freqs)
    assert.are.equal(64, freqs.numel)
  end)

  it("hz2bark", function()
    local bark = ins.signal.hz2bark(1000.0)
    assert.is_not_nil(bark)
    assert.is_true(bark > 0)
  end)

  it("bark2hz", function()
    local hz = ins.signal.bark2hz(10.0)
    assert.is_not_nil(hz)
    assert.is_true(hz > 0)
  end)

  it("bark_hz_roundtrip", function()
    local hz_orig = 1000.0
    local bark = ins.signal.hz2bark(hz_orig)
    local hz_back = ins.signal.bark2hz(bark)
    assert.is_true(math.abs(hz_back - hz_orig) < 10.0)
  end)
end)

File: tests/cuda/julia/test_signal_acoustics.jl
# Signal acoustics CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal_acoustics.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

function approx(a, b; atol=1e-5)
    return abs(a - b) < atol
end

# ============================================================================
# Acoustic Scale Conversions
# ============================================================================
println("=== Signal Acoustics ===")

# mel2hz
mel_arr = Insight.from_data([0.0, 1000.0, 2000.0])
mel_gpu = Insight.to_array(mel_arr, Insight.GPUPlace(0))
hz_arr = Insight.mel2hz(mel_gpu)
check("mel2hz", Insight.numel(hz_arr) == 3)

# hz2mel
hz_arr = Insight.from_data([0.0, 1000.0, 4000.0])
hz_gpu = Insight.to_array(hz_arr, Insight.GPUPlace(0))
mel_arr = Insight.hz2mel(hz_gpu)
check("hz2mel", Insight.numel(mel_arr) == 3)

# mel roundtrip
hz_orig = Insight.from_data([100.0, 500.0, 1000.0, 4000.0, 8000.0])
hz_gpu = Insight.to_array(hz_orig, Insight.GPUPlace(0))
mel = Insight.hz2mel(hz_gpu)
hz_back = Insight.mel2hz(mel)
orig_data = Insight.to_data(hz_orig)
back_data = Insight.to_data(hz_back)
check("mel_roundtrip", all(approx.(orig_data, back_data, atol=1.0)))

# mel_frequencies
freqs = Insight.mel_frequencies(128, f_low=0.0, f_high=11025.0)
check("mel_frequencies", Insight.numel(freqs) == 128)

# mel_frequencies custom
freqs = Insight.mel_frequencies(64, f_low=20.0, f_high=8000.0)
check("mel_frequencies_custom", Insight.numel(freqs) == 64)

# hz2bark
hz_arr = Insight.from_data([100.0, 500.0, 2000.0])
hz_gpu = Insight.to_array(hz_arr, Insight.GPUPlace(0))
bark_arr = Insight.hz2bark(hz_gpu)
check("hz2bark", Insight.numel(bark_arr) == 3)

# bark2hz
bark_arr = Insight.from_data([1.0, 5.0, 15.0])
bark_gpu = Insight.to_array(bark_arr, Insight.GPUPlace(0))
hz_arr = Insight.bark2hz(bark_gpu)
check("bark2hz", Insight.numel(hz_arr) == 3)

# bark roundtrip
hz_orig = Insight.from_data([100.0, 500.0, 2000.0, 6000.0])
hz_gpu = Insight.to_array(hz_orig, Insight.GPUPlace(0))
bark = Insight.hz2bark(hz_gpu)
hz_back = Insight.bark2hz(bark)
orig_data = Insight.to_data(hz_orig)
back_data = Insight.to_data(hz_back)
check("bark_roundtrip", all(approx.(orig_data, back_data, atol=10.0)))

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
